File: src/test/java/de/pflugradts/passbird/domain/model/transfer/OutputTest.java
package de.pflugradts.passbird.domain.model.transfer;

import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import static org.assertj.core.api.Assertions.assertThat;

class OutputTest {

    @Nested
    class InstantiationTest {

        @Test
        void shouldHaveBytes() {
            // given
            final var givenBytes = Bytes.bytesOf(new byte[]{1, 2, 3});
            final var givenOutput = Output.of(givenBytes);

            // when
            final var actual = givenOutput.getBytes();

            // then
            assertThat(actual).isNotNull().isEqualTo(givenBytes);
        }

        @Test
        void shouldInstantiateFromByteArray() {
            // given
            final var givenByteArray = new byte[]{1, 2, 3};
            final var expectedBytes = Bytes.bytesOf(givenByteArray);

            // when
            final var actual = Output.of(givenByteArray);

            // then
            assertThat(actual).isNotNull()
                    .extracting(Output::getBytes)
                    .isNotNull().isEqualTo(expectedBytes);
        }

        @Test
        void shouldInstantiateEmptyOutput() {
            // given / when / then
            assertThat(Output.empty()).isNotNull()
                    .extracting(Output::getBytes)
                    .isNotNull().isEqualTo(Bytes.emptyBytes());
        }

    }

}

File: src/test/kotlin/de/pflugradts/passbird/domain/model/transfer/OutputTest.kt
package de.pflugradts.passbird.domain.model.transfer

import de.pflugradts.passbird.domain.model.shell.Shell.Companion.emptyShell
import de.pflugradts.passbird.domain.model.shell.Shell.Companion.shellOf
import de.pflugradts.passbird.domain.model.transfer.Output.Companion.emptyOutput
import de.pflugradts.passbird.domain.model.transfer.Output.Companion.outputOf
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import strikt.api.expectThat
import strikt.assertions.isEqualTo
import strikt.assertions.isFalse
import strikt.assertions.isTrue

class OutputTest {
    @Nested
    inner class InstantiationTest {
        @Test
        fun `should have shell`() {
            // given
            val givenShell = shellOf(byteArrayOf(1, 2, 3))
            val givenOutput = outputOf(givenShell)

            // when
            val actual = givenOutput.shell

            // then
            expectThat(actual) isEqualTo givenShell
        }

        @Test
        fun `should instantiate from byte array`() {
            // given
            val givenByteArray = byteArrayOf(1, 2, 3)
            val expectedShell = shellOf(givenByteArray)

            // when
            val actual = outputOf(givenByteArray)

            // then
            expectThat(actual.shell) isEqualTo expectedShell
        }

        @Test
        fun `should instantiate empty output`() {
            // given / when / then
            expectThat(emptyOutput().shell) isEqualTo emptyShell()
        }
    }

    @Nested
    inner class EqualsTest {

        @Test
        fun `should be equal to itself`() {
            // given
            val output1 = outputOf(shellOf("abc"))
            val output2 = output1

            // when
            val actual = output1.equals(output2)

            // then
            expectThat(actual).isTrue()
        }

        @Test
        fun `should be equal to output with equal shell`() {
            // given
            val shell = shellOf("abc")
            val sameShell = shellOf("abc")
            val output1 = outputOf(shell)
            val output2 = outputOf(sameShell)

            // when
            val actual = output1.equals(output2)

            // then
            expectThat(actual).isTrue()
        }

        @Test
        fun `should not be equal to output with other shell`() {
            // given
            val shell = shellOf("abc")
            val otherShell = shellOf("abd")
            val output1 = outputOf(shell)
            val output2 = outputOf(otherShell)

            // when
            val actual = output1.equals(output2)

            // then
            expectThat(actual).isFalse()
        }

        @Test
        fun `should not be equal to other class`() {
            // given
            val shell = shellOf("abc")
            val output = outputOf(shell)

            // when
            val actual = output.equals(shell)

            // then
            expectThat(actual).isFalse()
        }

        @Test
        fun `should not be equal to null`() {
            // given
            val output = outputOf(shellOf("abc"))

            // when
            val actual = output.equals(null)

            // then
            expectThat(actual).isFalse()
        }
    }
}
